2021-11-monthly-report.org

```diff
@@ -0,0 +1,413 @@
+
+** Planned Activities
+
+
+
+** Detailed changelog
+
+*** Refactor TyTy::ADTType
+
+In order to support enums, we could have implemented a new TyTy module for this, to then subsequently update the typechecking code. But in alot of ways enum's are just another type of algebraic data type which also has the side effect of canonicalizing the flow of how we work with these types instead of inventing new paths in the compiler. The change here was that the ADT Type was originally designed for unit-structs, structs and tuple structs. But really an enum is an ADT with multiple variants and struts, tuple structs are ADT's with a single variant and finally a unit-struct is one with no variants. This change was rather large to decouple and refactor ADT's into variants but has helped tackle some technical debt along the way. Thanks to [[https://github.com/flip1995][flip1995]] for point us in this direction.
+
+*** Add new rust_internal_error
+
+This new API is designed to distinguish between internal compiler errors and actual program errors. Assertions are good but somtimes you want to be able to provide extra contextural information.
+
+*** Handle forward declared items within blocks
+
+Rust allows for Items such as functions to be declared at the bottom of a BlockExpr which can be referenced at any point within that conxtext such as: https://godbolt.org/z/PGqnz1nve
+
+#+BEGIN_SRC rust
+pub fn main() {
+    let a;
+    a = foo { a: 123, b: 456f32 };
+
+    let mut a = 123;
+    a = bar(a);
+
+    let mut b = 456f32;
+    b = bar(b);
+
+    let aa = X;
+
+    let bb:[i32; X];
+
+    fn bar<T>(x: T) -> T {
+        x
+    }
+
+    struct foo {
+        a: i32,
+        b: f32,
+    };
+
+    const X:usize = 2;
+}
+#+END_SRC
+
+*** Fix unhandled TypeBounds
+
+This test case qithin the rust testsuite https://github.com/rust-lang/rust/blob/d5a0c7cb036032288a4a5443b54ba061ec12ee26/src/test/ui/higher-rank-trait-bounds/hrtb-fn-like-trait-object.rs raised bugs with unhandled type bounds for TypeAlises:
+
+#+BEGIN_SRC rust
+type FnObject<'b> = dyn for<'a> FnLike<&'a isize, &'a isize> + 'b;
+#+END_SRC
+
+The bug here was that we had two ways in HIR to represent bounds here, and we had an opertunity to desugar our AST from two types:
+
+- AST::TraitObjectType
+- AST::TraitObjectTypeOneBound
+
+Into a single HIR::TraitObjectType
+
+*** Operator Overloading
+
+Rust supports operator overloading of many different operations, we have added support for all the regular arithmetic operators (+,-,*,/,%), compound assignments such as (+=, ...), the unary negation operators (!x, and -x). There is support for the deref operations but there are a few bugs to work though there to get it right. Since we cannot compile libcore yet we require you to define the lang items you want to use within your crate, we have taken the same traits from libcore to be sure that we can compile the correctly.
```

Comment:
```suggestion
Rust supports operator overloading of many different operations, we have added support for all the regular arithmetic operators (+,-,*,/,%), compound assignments such as (+=, ...), the unary negation operators (!x, and -x). There is support for the deref operations but there are a few bugs to work though there to get it right. Since we cannot compile libcore yet we require you to define the lang items you want to use within your crate, we have taken the same traits from libcore to be sure that we can compile them correctly.
```